osu.Game/Rulesets/Difficulty/Skills/Skill.cs

```diff
@@ -61,19 +68,41 @@ public abstract class Skill
         protected Skill(Mod[] mods)
         {
             this.mods = mods;
+            Previous = new ReverseQueue<DifficultyHitObject>(PreviousCollectionSoftCapacity + 1);
         }
 
         /// <summary>
         /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
         /// </summary>
         public void Process(DifficultyHitObject current)
         {
+            RemoveExtraneousHistory(current);
+
             CurrentStrain *= strainDecay(current.DeltaTime);
             CurrentStrain += StrainValueOf(current) * SkillMultiplier;
 
             currentSectionPeak = Math.Max(CurrentStrain, currentSectionPeak);
 
-            Previous.Push(current);
+            AddToHistory(current);
+        }
+
+        /// <summary>
+        /// Remove objects from <see cref="Previous"/> that are no longer needed for calculations from the current object onwards.
+        /// </summary>
+        /// <param name="current">The <see cref="DifficultyHitObject"/> to be processed.</param>
+        protected virtual void RemoveExtraneousHistory(DifficultyHitObject current)
```

Comment:
I'm not really a fan of this, do you have a situation in which the number of elements in `Previous` changes dynamically? Especially being virtual and all (along with `AddToHistory`).

Reply:
Yes, the PP+ algorithm I'm porting has a skill that requires time based history retention, instead of a fixed number.